File: python/main_fergus2003.py
#!/usr/bin/env python3
"""
main_fergus2003.py — Driver script for the Fergus et al. (2003) reimplementation.

Directory layout:
    data/
      motorbikes/    (*.jpg / *.png)
      airplanes/
      faces/
      cars_rear/
      background/

Usage:
    python main_fergus2003.py
    python main_fergus2003.py --category faces --parts 6 --iter 80
"""
from __future__ import annotations

import argparse
import logging
import pickle
import random
from pathlib import Path
from typing import Dict, List, Tuple

import matplotlib.pyplot as plt
import numpy as np
from PIL import Image
from sklearn.metrics import roc_curve

from constellation_model import (
    ConstellationModel,
    ConstellationRecognizer,
    EMConstellationLearner,
)
from feature_extraction import FeatureExtractor, Features, load_images

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s  %(levelname)s  %(message)s",
    datefmt="%H:%M:%S",
)
log = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
def main():
    parser = argparse.ArgumentParser(description="Fergus et al. 2003 — constellation model")
    parser.add_argument("--data_root",  default="data",           help="root data directory")
    parser.add_argument("--categories", nargs="+",
                        default=["motorbikes", "airplanes", "faces", "cars_rear"])
    parser.add_argument("--parts",      type=int, default=6,      help="number of parts P")
    parser.add_argument("--iter",       type=int, default=60,     help="EM iterations")
    parser.add_argument("--n_feat",     type=int, default=30,     help="features per image")
    parser.add_argument("--k_pca",      type=int, default=15,     help="PCA dimensions")
    parser.add_argument("--patch",      type=int, default=11,     help="patch size (pixels)")
    parser.add_argument("--s_min",      type=int, default=4,      help="min detector scale")
    parser.add_argument("--s_max",      type=int, default=40,     help="max detector scale")
    parser.add_argument("--seed",       type=int, default=42)
    parser.add_argument("--save_models", action="store_true",     help="pickle learnt models")
    args = parser.parse_args()

    rng = np.random.default_rng(args.seed)
    random.seed(args.seed)

    data_root = Path(args.data_root)
    bg_dir    = data_root / "background"

    # ---- Background images -----------------------------------------------
    log.info("Loading background images from %s", bg_dir)
    bg_paths = sorted(list(bg_dir.glob("*.jpg")) + list(bg_dir.glob("*.png")))
    bg_imgs  = load_images(bg_paths)
    log.info("  %d background images loaded.", len(bg_imgs))

    results: Dict[str, dict] = {}

    for category in args.categories:
        log.info("\n" + "=" * 55)
        log.info("Category: %s", category)
        log.info("=" * 55)

        cat_dir = data_root / category
        paths   = sorted(list(cat_dir.glob("*.jpg")) + list(cat_dir.glob("*.png")))
        idx     = list(range(len(paths)))
        rng.shuffle(idx)
        n_train = len(idx) // 2
        train_paths = [paths[i] for i in idx[:n_train]]
        test_paths  = [paths[i] for i in idx[n_train:]]
        log.info("  train=%d  test=%d", len(train_paths), len(test_paths))

        # ---- Feature extraction ----------------------------------------
        log.info("  Extracting training features...")
        extractor = FeatureExtractor(
            n_feat=args.n_feat, patch_sz=args.patch,
            k_pca=args.k_pca, s_min=args.s_min, s_max=args.s_max,
        )
        train_imgs = load_images(train_paths)
        train_feats = extractor.fit_transform(train_imgs)

        bg_feats = extractor.transform(bg_imgs)

        # ---- EM learning -----------------------------------------------
        log.info("  Learning constellation model (P=%d, %d EM iters)...",
                 args.parts, args.iter)
        learner = EMConstellationLearner(P=args.parts, n_iter=args.iter, verbose=True)
        model   = learner.fit(train_feats, bg_feats)

        if args.save_models:
            out_path = Path(f"model_{category}.pkl")
            with open(out_path, "wb") as f:
                pickle.dump(model, f)
            log.info("  Saved model → %s", out_path)

        # ---- Recognition -----------------------------------------------
        log.info("  Recognizing test images...")
        recognizer = ConstellationRecognizer(model)

        test_imgs  = load_images(test_paths)
        test_feats = extractor.transform(test_imgs)
        R_fg = recognizer.predict(test_feats)

        n_bg_test = len(test_paths)
        bg_test_paths = random.sample(bg_paths, min(n_bg_test, len(bg_paths)))
        bg_test_imgs  = load_images(bg_test_paths)
        bg_test_feats = extractor.transform(bg_test_imgs)
        R_bg = recognizer.predict(bg_test_feats)

        # ---- Evaluation ------------------------------------------------
        scores = np.concatenate([R_fg, R_bg])
        labels = np.concatenate([np.ones(len(R_fg)), np.zeros(len(R_bg))])
        eer    = compute_eer(scores, labels)
        acc    = (1 - eer) * 100

        log.info("  ROC EER = %.1f%%  |  Accuracy ≈ %.1f%%", eer * 100, acc)

        results[category] = {
            "model": model, "extractor": extractor,
            "R_fg": R_fg, "R_bg": R_bg,
            "eer": eer, "acc": acc,
        }

        # ---- Shape model visualisation ---------------------------------
        plot_shape_model(model, category)
        plt.savefig(f"shape_model_{category}.png", dpi=150, bbox_inches="tight")
        plt.close()

    # ---- Summary ---------------------------------------------------------
    print("\n" + "=" * 45)
    print(f"{'Dataset':<20}  {'Accuracy (%)':>12}")
    print("-" * 35)
    for cat, r in results.items():
        print(f"{cat:<20}  {r['acc']:>12.1f}")
    print("=" * 45)

    plot_roc(results)
    plt.savefig("roc_curves.png", dpi=150, bbox_inches="tight")
    plt.close()
    log.info("Saved roc_curves.png")


# ---------------------------------------------------------------------------
def compute_eer(scores: np.ndarray, labels: np.ndarray) -> float:
    fpr, tpr, _ = roc_curve(labels, scores)
    fnr = 1 - tpr
    idx = np.argmin(np.abs(fpr - fnr))
    return float((fpr[idx] + fnr[idx]) / 2)


def plot_shape_model(model: ConstellationModel, title: str):
    P      = model.P
    mu     = model.mu
    Sigma  = model.Sigma
    fig, ax = plt.subplots(1, 1, figsize=(5, 5))
    colors = plt.cm.tab10(np.linspace(0, 1, P))
    for p in range(P):
        mx = mu[2 * p];     my = mu[2 * p + 1]
        sx = np.sqrt(Sigma[2*p,   2*p])
        sy = np.sqrt(Sigma[2*p+1, 2*p+1])
        theta = np.linspace(0, 2 * np.pi, 100)
        ax.plot(mx + sx * np.cos(theta), my + sy * np.sin(theta),
                color=colors[p], linewidth=2)
        ax.plot(mx, my, "o", color=colors[p], markersize=8)
        ax.text(mx + 0.01, my + 0.01, f"P{p+1}", color=colors[p], fontsize=10)
    ax.set_aspect("equal")
    ax.grid(True, alpha=0.3)
    ax.set_title(f"{title} — shape model (P={P})")
    ax.set_xlabel("x (normalised)")
    ax.set_ylabel("y (normalised)")


def plot_roc(results: dict):
    fig, ax = plt.subplots(figsize=(6, 5))
    from sklearn.metrics import roc_curve as _roc
    for cat, r in results.items():
        scores = np.concatenate([r["R_fg"], r["R_bg"]])
        labels = np.concatenate([np.ones(len(r["R_fg"])), np.zeros(len(r["R_bg"]))])
        fpr, tpr, _ = _roc(labels, scores)
        ax.plot(fpr, tpr, label=f"{cat} (acc={r['acc']:.1f}%)")
    ax.plot([0, 1], [0, 1], "k--", alpha=0.4)
    ax.set_xlabel("False Positive Rate")
    ax.set_ylabel("True Positive Rate")
    ax.set_title("ROC Curves — Fergus et al. (2003)")
    ax.legend()
    ax.grid(True, alpha=0.3)


# ---------------------------------------------------------------------------
if __name__ == "__main__":
    main()

File: matlab/main_fergus2003.m
%% main_fergus2003.m
% Driver script for the Fergus et al. (2003) constellation model.
%
% Directory layout expected:
%   data/
%     motorbikes/   (*.jpg or *.png)
%     airplanes/
%     faces/
%     cars_rear/
%     background/
%
% Usage: run this script from the repo root.
%
% Nikita Letov, McGill ECSE 626 reimplementation

clear; clc; close all;
rng(42);  % reproducibility

%% ---- Settings ----------------------------------------------------------
DATA_ROOT  = fullfile('data');
CATEGORIES = {'motorbikes', 'airplanes', 'faces', 'cars_rear'};
BG_DIR     = fullfile(DATA_ROOT, 'background');

P          = 6;        % number of parts
N_FEAT     = 30;       % max features per image
K_PCA      = 15;       % PCA dimensions
PATCH_SZ   = 11;       % patch size in pixels
S_MIN      = 4;        % min detector scale
S_MAX      = 40;       % max detector scale
N_ITER     = 60;       % EM iterations

%% ---- Load background images -------------------------------------------
fprintf('=== Loading background images ===\n');
bg_files = [dir(fullfile(BG_DIR,'*.jpg')); dir(fullfile(BG_DIR,'*.png'))];
bg_imgs  = load_images({bg_files.name}, BG_DIR);

%% ---- Compute background PCA (will be reused across categories) ---------
fprintf('=== Background feature extraction ===\n');
[~, ~, A_bg, pca_bg] = extract_features(bg_imgs, N_FEAT, PATCH_SZ, K_PCA, S_MIN, S_MAX);

%% ---- Per-category train / test -----------------------------------------
results = struct();

for ci = 1:length(CATEGORIES)
    cat = CATEGORIES{ci};
    fprintf('\n========== Category: %s ==========\n', cat);

    % Load images
    cat_dir  = fullfile(DATA_ROOT, cat);
    files    = [dir(fullfile(cat_dir,'*.jpg')); dir(fullfile(cat_dir,'*.png'))];
    n_total  = length(files);
    idx_all  = randperm(n_total);
    n_train  = floor(n_total / 2);

    train_files = {files(idx_all(1:n_train)).name};
    test_files  = {files(idx_all(n_train+1:end)).name};

    fprintf('Train: %d  |  Test: %d\n', length(train_files), length(test_files));

    % ---- Feature extraction (training) ----------------------------------
    fprintf('  Extracting training features...\n');
    train_imgs = load_images(train_files, cat_dir);
    [X_tr, S_tr, A_tr, pca_model] = extract_features( ...
        train_imgs, N_FEAT, PATCH_SZ, K_PCA, S_MIN, S_MAX);

    % ---- EM learning ----------------------------------------------------
    fprintf('  Running EM learning...\n');
    model = learn_constellation(X_tr, S_tr, A_tr, bg_imgs, A_bg, P, N_ITER, true);
    model.pca = pca_model;

    % ---- Recognition on test set ----------------------------------------
    fprintf('  Recognizing test images...\n');
    test_imgs = load_images(test_files, cat_dir);
    [X_te, S_te, A_te] = extract_features( ...
        test_imgs, N_FEAT, PATCH_SZ, K_PCA, S_MIN, S_MAX, pca_model);

    % Background test images
    n_bg_test = length(test_files);
    bg_test_idx = randperm(length(bg_files), min(n_bg_test, length(bg_files)));
    bg_test_imgs = load_images({bg_files(bg_test_idx).name}, BG_DIR);
    [X_bg_te, S_bg_te, A_bg_te] = extract_features( ...
        bg_test_imgs, N_FEAT, PATCH_SZ, K_PCA, S_MIN, S_MAX, pca_model);

    % Compute likelihood ratios
    R_fg = zeros(length(test_imgs), 1);
    for i = 1:length(test_imgs)
        R_fg(i) = recognize(model, X_te{i}, S_te{i}, A_te{i});
    end

    R_bg = zeros(length(bg_test_imgs), 1);
    for i = 1:length(bg_test_imgs)
        R_bg(i) = recognize(model, X_bg_te{i}, S_bg_te{i}, A_bg_te{i});
    end

    % ROC equal error rate
    labels = [ones(length(R_fg),1); zeros(length(R_bg),1)];
    scores = [R_fg; R_bg];
    eer    = compute_eer(scores, labels);

    fprintf('  ROC Equal Error Rate: %.1f%%\n', eer * 100);

    results.(cat).model  = model;
    results.(cat).R_fg   = R_fg;
    results.(cat).R_bg   = R_bg;
    results.(cat).eer    = eer;

    % ---- Visualise shape model ------------------------------------------
    plot_shape_model(model, cat);
end

%% ---- Summary table -----------------------------------------------------
fprintf('\n\n===== RESULTS SUMMARY =====\n');
fprintf('%-20s  %s\n', 'Dataset', 'Accuracy (%)');
fprintf('%s\n', repmat('-',1,35));
for ci = 1:length(CATEGORIES)
    cat = CATEGORIES{ci};
    acc = (1 - results.(cat).eer) * 100;
    fprintf('%-20s  %.1f\n', cat, acc);
end


%% =========================================================================
%  Helper functions
%% =========================================================================

function imgs = load_images(fnames, dir_path)
imgs = cell(numel(fnames), 1);
for k = 1:numel(fnames)
    img = imread(fullfile(dir_path, fnames{k}));
    if size(img, 3) == 3
        img = rgb2gray(img);
    end
    imgs{k} = im2double(img);
end
end

function eer = compute_eer(scores, labels)
% Sweep threshold and find equal error rate
thresholds = unique(scores);
best = inf;
eer  = 0.5;
for t = thresholds'
    tp = sum(scores(labels==1) >= t);
    fn = sum(scores(labels==1) < t);
    fp = sum(scores(labels==0) >= t);
    tn = sum(scores(labels==0) < t);
    fpr = fp / max(fp + tn, 1);
    fnr = fn / max(fn + tp, 1);
    if abs(fpr - fnr) < best
        best = abs(fpr - fnr);
        eer  = (fpr + fnr) / 2;
    end
end
end

function plot_shape_model(model, title_str)
P  = model.P;
mu = model.mu;
Sigma = model.Sigma;

figure('Name', ['Shape model: ' title_str]);
hold on;
colors = lines(P);
for p = 1:P
    mx = mu(2*p - 1);
    my = mu(2*p);
    sx = sqrt(Sigma(2*p-1, 2*p-1));
    sy = sqrt(Sigma(2*p,   2*p));
    draw_ellipse(mx, my, sx, sy, colors(p,:));
    plot(mx, my, 'o', 'Color', colors(p,:), 'MarkerFaceColor', colors(p,:));
    text(mx + 0.01, my + 0.01, sprintf('P%d', p), 'Color', colors(p,:));
end
axis equal; grid on;
title([title_str ' shape model (P=' num2str(P) ')']);
xlabel('x (normalised)'); ylabel('y (normalised)');
hold off;
end

function draw_ellipse(cx, cy, rx, ry, color)
theta = linspace(0, 2*pi, 100);
x = cx + rx * cos(theta);
y = cy + ry * sin(theta);
plot(x, y, '-', 'Color', color, 'LineWidth', 1.5);
end
